src/ghga_service_commons/http/correlation.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2021 - 2024 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tools to enhance traceability of HTTP requests across microservices."""
+
+from functools import partial
+from typing import Union
+
+import httpx
+from hexkit.correlation import (
+    CorrelationIdContextError,
+    get_correlation_id,
+    new_correlation_id,
+)
+
+CORRELATION_ID_HEADER_NAME = "X-Request-Id"
+
+__all__ = ["attach_correlation_id_to_requests"]
+
+
+def _cid_request_hook(request, generate_correlation_id: bool):
+    """Include the correlation ID in the request header so it is propagated.
+
+    If the correlation ID isn't set, one will be generated if `generate_correlation_id`
+    is set. Otherwise, an error will be raised.
+
+    Raises:
+        CorrelationIdContextError: when the correlation ID ContextVar is not set.
+        InvalidCorrelationIdError: when the correlation ID is invalid.
+    """
+    try:
+        correlation_id = get_correlation_id()
+    except CorrelationIdContextError:
+        if generate_correlation_id:
+            correlation_id = new_correlation_id()
+        else:
+            raise
+    request.headers[CORRELATION_ID_HEADER_NAME] = correlation_id
+
+
+async def _cid_request_hook_async(request, generate_correlation_id: bool):
+    """Async version of `_cid_request_hook`"""
+    _cid_request_hook(request, generate_correlation_id)
+
+
+def attach_correlation_id_to_requests(
+    client: Union[httpx.Client, httpx.AsyncClient],
+    *,
+    generate_correlation_id: bool,
+):
+    """Add an event hook to an httpx Client that includes the correlation ID header."""
```

Comment:
My idea was to provide a custom version of the httpx.(Async)Client that would already have the hook attached. Your solution is a bit more flexible in that it also allows other compatible Clients like the hishel.CacheClient.

Maybe we can have a 

```python
T = TypeVar('T', bound=Union[httpx.Client, httpx.AsyncClient])

def create_httpx_client(
    client_class: Type[T] = httpx.Client,
    *,
    generate_correlation_id: bool = False,
    **kwargs
) -> T:
    """Factory function to create an httpx client with a hook to add the correlation ID header."""
    client = client_class(**kwargs)
    event_hook = ...
    client.event_hooks.setdefault("request", []).append(event_hook)
    return client
```

Or maybe two functions `create_httpx_client` and `create_httpx_async_client` with corresponding default base class to avoid the TypeVar and the case distinction inside.

What do you think?

Reply:
It's almost the same form factor as the current approach, with the difference being that you no longer have the intellisense for instantiating the client. You can subclass `httpx.AsyncClient`, but then you are locked into that class type and have to choose between the flexibility offered via `generate_correlation_id` param in `__init__` _or_ intellisense (as modifying `__init__` obscures the parent's signature for intellisense.
 
I think the current function that takes a client instance as a parameter is a good balance. Another thought: we could add an `AsyncClient` and `StrictAsyncClient`, which add the event hook with the generation flag set to `True` and `False`, respectively.

Reply:
Let's just add AsyncClient.